### ISSUE TITLE
Ltac2: Could the return type of `fail` be `'a` instead of `unit`?

### ISSUE THREAD

**Issue by samuelgruetter** [coq/coq#17463]
In Ltac1, I can write

```coq
Ltac pred n :=
  lazymatch n with
  | O => fail
  | S ?m => m
  end.
```

In Ltac2, however, this leads to a typing error

```coq
Require Import Ltac2.Ltac2.

Ltac2 pred(n: constr): constr :=
  lazy_match! n with
  | O => fail
  | S ?m => m
  end.
```

because `fail` is defined as a notation that expands to `fail0 ()`, which does a `Control.enter` (whose return type is unit):

https://github.com/coq/coq/blob/7c4e8b7f54054d248375e0f3c367a540673cff3d/user-contrib/Ltac2/Notations.v#L63-L65

So, currently, `fail0` has type `unit -> unit`, but I would find it more useful if it had type `unit -> 'a`. Would this be possible?

The easiest way I see would simply be to delete the `Control.enter`. Or is there a really good reason it needs to be there?

**Comment by SkySkimmer** [comment]
Ltac1 fail is also focusing (ie does a enter), that is why there is gfail.
Why not use Control.zero directly? I guess you could also define ltac2 gfail (using Control.zero)

**Comment by ppedrot** [comment]
Yes.

**Comment by JasonGross** [comment]
> So, currently, `fail0` has type `unit -> unit`, but I would find it more useful if it had type `unit -> 'a`. Would this be possible?

I think you could also do the "prove false" hack:
```coq
Ltac2 rec fail0 (_ : unit) : 'a := Control.enter (fun _ => Control.zero (Tactic_failure None)); fail0 ().
```
(This might mess with backtracking, though, I'm not completely sure.)

**Comment by SkySkimmer** [comment]
That will loop if there are no goals. Also why would you do that instead of just Control.zero?

**Comment by samuelgruetter** [comment]
It would also be nice to have a formatted `fail` that takes printf-style arguments. Without understanding the meaning of the four type arguments of the Ltac2 type `format` (maybe I should read [this one](http://gallium.inria.fr/blog/format6/) more carefully, or is there Ltac2-specific documentation available?), I came up with this:

```coq
Require Import Ltac2.Ltac2.
Require Import Ltac2.Printf.

Ltac2 fail_with_message (msg : message) :=
  Control.enter (fun _ => Control.zero (Tactic_failure (Some msg))).

Ltac2 failf0 (fmt : ('a, unit, message, unit) format) : 'a :=
  Message.Format.kfprintf fail_with_message fmt.

Ltac2 Notation "failf" fmt(format) := failf0 fmt.

Ltac2 gfail_with_message (msg : message) :=
  Control.zero (Tactic_failure (Some msg)).

Ltac2 gfailf0 (fmt : ('a, unit, message, 'b) format) : 'a :=
  Message.Format.kfprintf gfail_with_message fmt.

Ltac2 Notation "gfailf" fmt(format) := gfailf0 fmt.

Fail Ltac2 test_nonunit_failf () :=
  if true then failf "nope" else constr:(1).

Ltac2 test_nonunit_gfailf () :=
  if true then gfailf "nope" else constr:(1).
```

It seems to work, but with `failf`, the same issue as with `fail` above comes up: If `fail` has to do `Control.enter`, it can't return `'a`, but only `unit`, which makes it less useful.

So, my question remains: 
Is there a good reason for Ltac2's `fail` to call `Control.enter` (except for "let's do the same as Ltac1 does")?

**Comment by JasonGross** [comment]
> Is there a good reason for Ltac2's `fail` to call `Control.enter` (except for "let's do the same as Ltac1 does")?

Sometimes it's useful to have `fail` mean "fail if any goals remain"?

**Comment by samuelgruetter** [comment]
> Sometimes it's useful to have `fail` mean "fail if any goals remain"?

Good point! But does that mean `fail` has to auto-focus? In Ltac1, yes, because there's no explicit focus command in Ltac1. But in Ltac2, I could also imagine writing `tac1; Control.enter (fun _ => fail)`. But maybe there's already too much Ltac2 code out there that does `tac1; fail`, or too many people who expect `fail` to auto-focus? In that case, would there be any other means to allow `fail` to return `'a` instead of `unit`?

**Comment by SkySkimmer** [comment]
I don't think we want a difference between ltac1 and ltac2 for this, and there's no way to return `'a` for something that doesn't always fail.

**Comment by samuelgruetter** [comment]
> I don't think we want a difference between ltac1 and ltac2 for this, and there's no way to return `'a` for something that doesn't always fail.

I think that's a good way of summarizing the design constraints :+1: Saying that `fail` doesn't always fail sounds funny, but indeed, if an auto-focusing `fail` is run on zero goals, it succeeds, so it has to return something (such as unit), and can't just make up an arbitrary `'a`.

So I guess I'd have to use a `gfail` for this case, which could be defined as `Control.zero (Tactic_failure None)`. That sounds all fine, except that thinking ahead about naming, I might want to propose variants of `fail` and `gfail` that take `printf`-style arguments, which then probably would have to be named `failf` and `gfailf`. I guess `failf` might be acceptable, but `gfailf` is a bit too idiosyncratic... As in, I imagine myself trying to sell Ltac2 as a nice concise language to describe domain-specific automated reasoning steps, and the first thing that jumps into my audience's eyes is a `gfailf`? Too high [wtf/min rate](https://i.redd.it/0ixihi8o69a31.jpg)... :sweat_smile: 
Any opinions on naming here?

**Comment by SkySkimmer** [comment]
Do you actually need the `f` suffix? With notations you can have both 0-argument gfail and formatted gfail:
~~~coq
Require Import Ltac2.Ltac2.

Ltac2 gfail0 () := Control.zero (Tactic_failure None).

Ltac2 gfail fmt := Message.Format.kfprintf (fun msg => Control.zero (Tactic_failure (Some msg))) fmt.

Ltac2 Notation "gfail" fmt(format) := gfail fmt.
Ltac2 Notation "gfail" := (gfail0 ()).

Goal True.
  Fail gfail. (* error uncaught Tactic_failure None *)
  Fail gfail "hello". (* error uncaught Tactic_failure Some message:"hello" *)
  Fail gfail "hello %s" "world". (* error uncaught Tactic_failure Some message:"hello world" *)
~~~

**Comment by samuelgruetter** [comment]
Oh that's pretty cool!
It would also be good to allow `printf`-style arguments for `fail`, but currently the `fail` Notation is defined without quotes:
```coq
Ltac2 Notation fail := fail0 ().
```
and it seems that to allow `printf`-style arguments for `fail`, the no-argument version would need quotes as well:
```coq
Ltac2 Notation "fail" := fail0 ().
```
Do you think adding these quotes would have any undesired side effects?

**Comment by SkySkimmer** [comment]
Using the quotes means the qualified `Ltac2.Notations.fail` doesn't exist
I guess we can just define all 3 if we want
~~~coq
Ltac2 gfail0 () := Control.zero (Tactic_failure None).

Ltac2 gfail_ fmt := Message.Format.kfprintf (fun msg => Control.zero (Tactic_failure (Some msg))) fmt.

Ltac2 Notation gfail := (gfail0 ()).
Ltac2 Notation "gfail" := (gfail0 ()).
Ltac2 Notation "gfail" fmt(format) := gfail_ fmt.
~~~

**Comment by samuelgruetter** [comment]
Ah cool, so I think I'm getting close to feeling able to submit a PR for a `gfail` and `fail` with and without `printf`-style arguments.

However, the pattern "Sam wants a value-returning version of an existing unit-returning tactic" already starts repeating itself, namely for `first`. I'd like to have the following:

```coq
Require Import Ltac2.Ltac2.

Ltac2 Type exn ::= [ No_applicable_tactic ].

Ltac2 rec gfirst0 tacs :=
  match tacs with
  | [] => Control.zero No_applicable_tactic
  | tac :: tacs => orelse tac (fun _ => gfirst0 tacs)
  end.

Ltac2 Notation "gfirst" "[" tacs(list0(thunk(tactic(6)), "|")) "]" := gfirst0 tacs.

Ltac2 pred n :=
  lazy_match! n with
  | O => Control.zero (Tactic_failure None)
  | S ?m => m
  end.

Ltac2 succ n := constr:(S $n).

Ltac2 test n := gfirst [ pred n | succ n ].
```

Similarly to `fail`/`gfail`, it would almost be possible to use the same tactic for the unit-returning and value-returning case, except that people expect the unit-returning variant to auto-focus on the goal, so we need to define a `g` variant that does the same as the original variant except that the `Control.enter` is removed.

In the case of `gfirst`, I'm not sure if `g` is the best prefix, because there, the emphasis is more on returning a value vs not returning a value, so I would find `vfirst` or `first_val` a better name. But it would also be desirable to be consistent with the naming of `gfail`. So maybe `gfail` should actually be `vfail`? Or any other naming ideas/opinions?

**Comment by ppedrot** [comment]
> Using the quotes means the qualified Ltac2.Notations.fail doesn't exist

It also means we add a parsing rules, and parsing rules are highly non-modular. I precisely wrote the code that way to reduce the Ltac1 catastrophe where everything is a notation.

**Comment by SkySkimmer** [comment]
So what shall we do about this?

**Comment by ppedrot** [comment]
I'm fine with a notation guarded by an import, just like we do for printf.

**Comment by samuelgruetter** [comment]
A few months ago, based on this discussion, I wrote [this file](https://github.com/mit-plv/coqutil/blob/c1f3e55ae86b78bfed856acbc07779bb0975f027/src/coqutil/Ltac2Lib/Failf.v). Should I turn this into a PR?

**Comment by SkySkimmer** [comment]
The gfail part yes.

The fail part I'm not sure, with that implementation 
~~~coq
Goal forall (x: nat), True.
  intros.
  exact I; fail "The problem is %t" '(x + x).
~~~
errors (`x not found`) which I'm not sure is desired.

**Comment by SkySkimmer** [comment]
I think we would need to provide a primitive `wrapped_kfprintf : ((unit -> unit) -> unit) -> type of kfprintf` which calls the extra argument around the printing part of the application.
This still would not allow to write `fail "%t" 'x` (that would require notation changes) but would at least allow `fail "%a" (fun () c => fprintf "%t" (c ())) (fun () => 'x)` (ie use a thunked constr).
That's not fully satisfying though.
